crates/scsi/src/scsi/emulation/command.rs

```diff
@@ -0,0 +1,579 @@
+            0xb8 => Ok(Self::FormatPresets),
+            _ => Err(()),
+        }
+    }
+}
+
+impl From<VpdPage> for u8 {
+    fn from(pc: VpdPage) -> Self {
+        match pc {
+            VpdPage::Ascii(val) => val,
+            VpdPage::Ata => 0x89,
+            VpdPage::BlockDeviceCharacteristics => 0xb1,
+            VpdPage::BlockDeviceCharacteristicsExt => 0xb5,
+            VpdPage::BlockLimits => 0xb0,
+            VpdPage::BlockLimitsExt => 0xb7,
+            VpdPage::CfaProfile => 0x8c,
+            VpdPage::DeviceConstituents => 0x8b,
+            VpdPage::DeviceIdentification => 0x83,
+            VpdPage::ExtendedInquiry => 0x86,
+            VpdPage::FormatPresets => 0xb8,
+            VpdPage::LogicalBlockProvisioning => 0xb2,
+            VpdPage::ManagementNetworkAddresses => 0x85,
+            VpdPage::ModePagePolicy => 0x87,
+            VpdPage::PowerCondition => 0x8a,
+            VpdPage::PowerConsumption => 0x8d,
+            VpdPage::PortocolSpecificLogicalUnit => 0x90,
+            VpdPage::ProtocolSpecificPort => 0x91,
+            VpdPage::Referrals => 0xb3,
+            VpdPage::ScsiFeatureSets => 0x92,
+            VpdPage::ScsiPorts => 0x88,
+            VpdPage::SoftwareInterfaceIdentification => 0x84,
+            VpdPage::SupportedVpdPages => 0x00,
+            VpdPage::ThirdPartyCopy => 0x8f,
+            VpdPage::UnitSerialNumber => 0x80,
+            VpdPage::ZonedBlockDeviceCharacteristics => 0xb6,
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum SenseFormat {
+    Fixed,
+    Descriptor,
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum ModePageSelection {
+    AllPageZeros,
+    Single(ModePage),
+}
+
+#[derive(Debug)]
+pub enum LunIndependentCommand {
+    ReportLuns(ReportLunsSelectReport),
+}
+
+#[derive(Debug)]
+pub enum LunSpecificCommand {
+    Inquiry(Option<VpdPage>),
+    ModeSense6 {
+        pc: ModeSensePageControl,
+        mode_page: ModePageSelection,
+        /// Disable block descriptors
+        dbd: bool,
+    },
+    Read10 {
+        /// Disable page out (i.e. hint that this page won't be accessed again
+        /// soon, so we shouldn't bother caching it)
+        dpo: bool,
+        /// Force unit access (i.e. bypass cache)
+        fua: bool,
+        lba: u32,
+        /// Group number, used for serveral SCSI features we don't implement
+        /// (such as tagging commands from different workloads for performance
+        /// monitoring purposes)
+        group_number: u8,
+        transfer_length: u16,
+    },
+    ReadCapacity10,
+    ReadCapacity16,
+    ReportSupportedOperationCodes {
+        /// SCSI RCTD bit: whether we should include timeout descriptors.
+        rctd: bool,
+        mode: ReportSupportedOpCodesMode,
+    },
+    RequestSense(SenseFormat),
+    TestUnitReady,
+}
+
+#[derive(Debug)]
+pub enum Command {
+    LunIndependentCommand(LunIndependentCommand),
+    LunSpecificCommand(LunSpecificCommand),
```

Comment:
It's worth noting that several of the "LUN-specific" commands here are used for multiple LUN types, not just block devices. (TEST UNIT READY, REQUEST SENSE, REPORT SUPPORTED OPERATION CODES, and INQUIRY are required for all LUN types; commands like READ CAPACITY and READ are used by several storage-like devices.)

So if we ever wanted to do more than just block devices, we'd need to do some more thinking about how we want to handle that. Unfortunately we can't just parse a command without knowing the intended device type, as sometimes operation codes are reused.

Anyway, cross that bridge when we get to it - just potentially useful background.